addons/udes_report/models/stock_export.py

```diff
@@ -62,79 +62,103 @@ def run_stock_file_export(self, context=None, send_file_via='user'):
-        by_location = lambda x: x.location_id.id
-        by_package_name = lambda x: x.package_id.name or ''
 
-        get_quants_by_loc = lambda quant_union: {
-            Location.browse(loc): Quant.union(*_quants)
-            for loc, _quants in groupby(quant_union.sorted(key=by_location),
-                                        key=by_location)}
+        def by_prod(record):
+            return record.product_id.id
+
+        def by_location(record):
+            return record.location_id.id
+
+        def by_package_name(record):
+            return record.package_id.name or ''
+
+        def by_lot_name(record):
+            return record.lot_id.name or ''
+
+        def get_quants_by_loc(quant_union):
+            return {Location.browse(loc): Quant.union(*_quants)
+                    for loc, _quants in groupby(quant_union.sorted(key=by_location),
+                                                key=by_location)}
+
+        locations = self.included_locations - self.excluded_locations
+        if not locations:
+            raise UserError(_("The specified list of Stock Locations is empty."))
+
+        all_quants = Quant.search([('location_id', 'child_of', locations.ids)])
 
         quants_by_prod_by_loc = OrderedDict()
         prod_summary = defaultdict(lambda: {'n_pkgs': 0, 'qty': 0})
 
-        for prod, _quants in groupby(all_quants.sorted(key=by_prod),
-                                     key=by_prod):
+        # Determine if lots/serials should be shown on the report
+        show_tracking = bool(Product.search([("tracking", "!=", "none")]))
+
+        for prod, _quants in groupby(all_quants.sorted(key=by_prod), key=by_prod):
             quant_prods = Product.browse(prod)
             quants_u = Quant.union(*_quants)
             quants_by_prod_by_loc[quant_prods] = get_quants_by_loc(quants_u)
 
-        def get_prod_qty(quants):
-            return sum(Quant.union(*quants).mapped('quantity'))
-
-        _logger.info(_('Creating Excel file'))
-        file_name = 'warehouse_stock_{TIMESTAMP}.xls'.format(
-            TIMESTAMP=datetime.now())
-        wb = xlwt.Workbook()
-
-        # 1/2) 'Stock File': one entry for each (prod, package) pair
+        stock_file_data = {"name": 'Stock File'}
 
-        stock_sheet = _create_sheet(wb, 'Stock File', ['Part Number',
-                                                       'Location',
-                                                       'Package',
-                                                       'Quantity'])
+        if show_tracking:
+            stock_file_data["column_titles"] = ['Part Number',
+                                                'Location',
+                                                'Package',
+                                                'Lot/Serial Number',
+                                                'Quantity']
+        else:
+            stock_file_data["column_titles"] = ['Part Number', 'Location', 'Package', 'Quantity']
```

Comment:
Can we split this out to avoid repeating common columns? If we want to maintain column order:
```
headers = ['Part Number', 'Location', 'Package']

if show_tracking:
   headers.append('Lot/Serial Number')

headers.append('Quantity')

stock_file_data["column_titles"] = headers
```

Reply:
alternatively you could use insert to add in the lot column

Reply:
Changed as suggested